cf-file-scanner/src/main/java/org/broadinstitute/ddp/cf/FileScanner.java

```diff
@@ -0,0 +1,280 @@
+        CommandResult result = runCommand(false, null,
+                FRESHCLAM_BIN,
+                "--config-file=" + FRESHCLAM_CONF,
+                "--datadir=" + DB_DIR);
+        logger.info("Output: " + result.getOutput());
+        logger.info("Status: " + result.getExitCode());
+
+        int status = result.getExitCode();
+        if (status == 0) {
+            dbUpdatedAt = Instant.now();
+            long elapsed = System.currentTimeMillis() - start;
+            logger.info("Finished refreshing the database at: " + dbUpdatedAt + " (took " + elapsed + "ms)");
+        } else {
+            throw new RuntimeException("Unable to refresh database, freshclam exited with status: " + status);
+        }
+    }
+
+    private static ScanResult runClamscan(byte[] content) {
+        CommandResult result = runCommand(true, content,
+                CLAMSCAN_BIN,
+                "--database=" + DB_DIR,
+                "-");   // Dash means to read from stdin.
+        logger.info("Output: " + result.getOutput());
+        logger.info("Status: " + result.getExitCode());
+
+        int status = result.getExitCode();
+        if (status == 0) {
+            return ScanResult.CLEAN;
+        } else if (status == 1) {
+            return ScanResult.INFECTED;
+        } else {
+            throw new RuntimeException("Error while scanning file, clamscan exited with status: " + status);
+        }
+    }
+
+    @Override
+    public void accept(Message message, Context context) {
+        logger.info("Received message: eventId=" + context.eventId() + " timestamp=" + context.timestamp());
+
+        String bucketName = message.getAttributes().get(ATTR_BUCKET_ID);
+        String fileName = message.getAttributes().get(ATTR_OBJECT_ID);
+        if (bucketName == null || bucketName.isBlank()) {
+            logger.severe("Bucket name is missing in message");
+            return;
+        } else if (fileName == null || fileName.isBlank()) {
+            logger.severe("Bucket filename is missing in message");
+            return;
+        }
+
+        BlobId blobId = BlobId.of(bucketName, fileName);
+        logger.info("Looking up file: " + blobId.toString());
+        Blob blob = storage.get(blobId);
+        if (blob == null || !blob.exists()) {
+            logger.severe("Could not find file: " + blobId.toString());
+            // This might be a duplicate (due to pubsub at-least-once delivery), in which case the
+            // file might have already been moved. Exit gracefully so we don't incur a cold start.
+            return;
+        }
+
+        Instant createdAt = Instant.ofEpochMilli(blob.getCreateTime());
+        logger.info("Found file that was created at: " + createdAt.toString());
+
+        // Ensure database is up-to-date before scanning.
+        checkDatabaseFiles();
+
+        logger.info("Scanning file: " + blobId.toString());
+        ScanResult result = runClamscan(blob.getContent());
```

Comment:
Can we stream the file contents instead of reading them all into memory? Would be surprised if clamav cannot take a file stream???

Reply:
Is there a way to get the contents of a bucket object as a stream? I didn't see anything when looking at the methods on the `Blob` Java class. Happy to stream more things if possible!

Reply:
Starting from a Blob, I think you can stream stuff using:
https://googleapis.dev/java/google-cloud-storage/latest/com/google/cloud/storage/Blob.html#reader-com.google.cloud.storage.Blob.BlobSourceOption...-
(I think you might be able to use directly or create an InputStream from it if treated as an nio Channel)

Reply:
Thanks! I did some digging around and saw the `reader()` thing as well. I also saw that gcloud libraries come with some nifty byte stream utils, so I stitched something together.

I did a quick comparison by running things locally with an 84MB file as input. Overall, I see that using streams does reduce memory usage but increases scanning time. I think this is because behind the scenes the `reader()` from the blob actually makes HTTP calls out to GCP to load chunks of bytes. So with streams there's going to be more roundtrips, roughly 10 seconds more on my local. It feels that memory is a more limited resource when it comes to Cloud Functions, so waiting a handful more seconds seems like an acceptable trade-off.

Before: There's a big spike starting around 6:32:12pm, which is when I think we start loading the blob contents.

![Screen Shot 2021-02-05 at 6 34 50 PM](https://user-images.githubusercontent.com/5713833/107101182-80709b80-67e4-11eb-9f98-eb19bf837e63.png)
<img width="789" alt="Screen Shot 2021-02-05 at 6 34 37 PM" src="https://user-images.githubusercontent.com/5713833/107101191-88c8d680-67e4-11eb-90d7-e5f7f3e7f539.png">

After: The spike is noticeable smaller and more flat, starting around 6:40:42pm.

![Screen Shot 2021-02-05 at 6 42 28 PM](https://user-images.githubusercontent.com/5713833/107101319-ebba6d80-67e4-11eb-8162-e9f2d986cf76.png)
<img width="826" alt="Screen Shot 2021-02-05 at 6 42 20 PM" src="https://user-images.githubusercontent.com/5713833/107101329-f543d580-67e4-11eb-90a7-7fa50f89b7e9.png">

Reply:
Yea if this substantially reduces memory usage, I'd be in favor of longer scan times tbh - as memory is unfortunately very limited on GCF, and not an "elastic" resource until time 😄